File: JatxMusicAndroid/musicTransmitterAndroid/src/main/java/jatx/musictransmitter/commons/Microphone.java
package jatx.musictransmitter.commons;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;

import jatx.debug.Debug;
import jatx.musiccommons.mp3.Frame;

/**
 * Created by jatx on 23.09.17.
 */

public class Microphone {
    private static int SAMPLE_RATE; // Hz
    private static int ENCODING;
    private static int CHANNEL_MASK;

    private static int BUFFER_SIZE;

    private static volatile byte[] data;
    private static volatile AudioRecord audioRecord;

    public static void start() throws MicrophoneInitException {
        audioRecord = findAudioRecord();
        if (audioRecord==null) {
            throw new MicrophoneInitException();
        }
        audioRecord.startRecording();
    }

    public static void stop() {
        //audioRecord.stop();
        audioRecord.release();
    }

    public static Frame readFrame(int position) throws MicrophoneReadException {
        try {
            int numBytesRead = audioRecord.read(data, 0, BUFFER_SIZE);
            return Frame.fromMicRawData(data, numBytesRead, position, SAMPLE_RATE);
        } catch (Throwable e) {
            throw new MicrophoneReadException(e);
        }
    }

    public static class MicrophoneReadException extends Exception {
        public MicrophoneReadException(Throwable e) {
            super(e);
        }
    }

    public static class MicrophoneInitException extends Exception {}

    private static AudioRecord findAudioRecord() {
        final int[] sampleRates = new int[] { 48000, 44100, 32000, 22050, 16000, 11025, 8000 };
        for (int rate : sampleRates) {
            for (short audioFormat : new short[] { AudioFormat.ENCODING_PCM_16BIT/*, AudioFormat.ENCODING_PCM_8BIT*/ }) {
                for (short channelConfig : new short[] { AudioFormat.CHANNEL_IN_STEREO/*, AudioFormat.CHANNEL_IN_MONO*/ }) {
                    try {
                        Log.e("Trying mic config", "Attempting rate " + rate + "Hz, bits: " + audioFormat + ", channel: "
                                + channelConfig);
                        BUFFER_SIZE = AudioRecord.getMinBufferSize(rate, channelConfig, audioFormat);

                        if (BUFFER_SIZE != AudioRecord.ERROR_BAD_VALUE) {
                            // check if we can instantiate and have a success
                            data = new byte[BUFFER_SIZE];
                            AudioRecord recorder = new AudioRecord(MediaRecorder.AudioSource.DEFAULT, rate, channelConfig, audioFormat, BUFFER_SIZE);

                            if (recorder.getState() == AudioRecord.STATE_INITIALIZED) {
                                Log.e("Attempt", "Success");
                                SAMPLE_RATE = rate;
                                ENCODING = audioFormat;
                                CHANNEL_MASK = channelConfig;
                                return recorder;
                            }
                        }
                    } catch (Exception e) {
                        Log.e("Attempt fault", rate + "Exception, keep trying.", e);
                    }
                }
            }
        }
        return null;
    }

}

File: JatxMusicAndroid2/musicTransmitterAndroid/src/main/java/jatx/musictransmitter/android/audio/Microphone.kt
package jatx.musictransmitter.android.audio

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import jatx.musiccommons.Frame
import jatx.musiccommons.frameFromMicRawData

object Microphone {
    private var sampleRate = 0
    private var bufferSize = 0
    private var encoding = 0
    private var channelMask = 0
    @Volatile private var data = ByteArray(0)
    @Volatile private var audioRecord: AudioRecord? = null

    @Throws(MicrophoneInitException::class)
    fun start() {
        if (audioRecord == null) audioRecord = findAudioRecord()
        audioRecord?.startRecording() ?: throw MicrophoneInitException()
    }

    fun stop() {
        audioRecord?.stop()
    }

    @Throws(MicrophoneReadException::class)
    fun readFrame(position: Int): Frame? {
        return try {
            val numBytesRead = audioRecord?.read(data, 0, bufferSize) ?: 0
            frameFromMicRawData(data, numBytesRead, position, sampleRate)
        } catch (e: Throwable) {
            throw MicrophoneReadException(e)
        }
    }

    private fun findAudioRecord(): AudioRecord? {
        val sampleRates =
            intArrayOf(48000, 44100, 32000, 22050, 16000, 11025, 8000)
        for (rate in sampleRates) {
            for (audioFormat in shortArrayOf(AudioFormat.ENCODING_PCM_16BIT.toShort())) {
                for (channelConfig in shortArrayOf(AudioFormat.CHANNEL_IN_STEREO.toShort() )) {
                    try {
                        println("Trying mic config: attempting rate $rate Hz, bits: $audioFormat, channel: $channelConfig")
                        bufferSize = AudioRecord.getMinBufferSize(rate, channelConfig.toInt(), audioFormat.toInt())
                        if (bufferSize != AudioRecord.ERROR_BAD_VALUE) { // check if we can instantiate and have a success
                            data = ByteArray(bufferSize)
                            val audioRecord = AudioRecord(
                                MediaRecorder.AudioSource.DEFAULT,
                                rate,
                                channelConfig.toInt(),
                                audioFormat.toInt(),
                                bufferSize
                            )
                            if (audioRecord.state == AudioRecord.STATE_INITIALIZED) {
                                println("Attempt success")
                                sampleRate = rate
                                encoding = audioFormat.toInt()
                                channelMask = channelConfig.toInt()
                                return audioRecord
                            }
                        }
                    } catch (e: Throwable) {
                        println("Find audio record attempt failed: $rate")
                    }
                }
            }
        }
        return null
    }

}

class MicrophoneReadException: Exception {
    constructor(e: Throwable): super(e)
}

class MicrophoneInitException: Exception()